File: paytmlibrary/src/main/java/com/bigfun/tm/HttpUtils.java
package com.bigfun.tm;

import android.app.Activity;
import android.os.Handler;
import android.os.Looper;
import android.os.SystemClock;
import android.text.TextUtils;
import android.util.Log;

import com.bigfun.tm.database.EventBean;
import com.bigfun.tm.database.EventManager;
import com.bigfun.tm.encrypt.EncryptUtil;
import com.bigfun.tm.model.LoginBean;
import com.bigfun.tm.model.PaymentOrderBean;
import com.google.gson.Gson;

import java.io.IOException;
import java.security.cert.CertificateException;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;

import javax.net.ssl.SSLContext;
import javax.net.ssl.SSLSocketFactory;
import javax.net.ssl.TrustManager;
import javax.net.ssl.X509TrustManager;

import okhttp3.Call;
import okhttp3.Callback;
import okhttp3.MediaType;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;

public class HttpUtils {

    private HttpUtils() {
        okHttpClient = getUnsafeOkHttpClient();
    }

    private static class InstanceHolder {
        private static HttpUtils instance = new HttpUtils();
    }

    public static HttpUtils getInstance() {
        return InstanceHolder.instance;
    }

    public MediaType mediaType = MediaType.parse("application/json; charset=utf-8");
    private static final long TIME_OUT = 30L;
    public Gson gson = new Gson();
    public OkHttpClient okHttpClient;
    //    public OkHttpClient okHttpClient = new OkHttpClient.Builder()
//            .connectTimeout(TIME_OUT, TimeUnit.SECONDS)
//            .writeTimeout(TIME_OUT, TimeUnit.SECONDS)
//            .readTimeout(TIME_OUT, TimeUnit.SECONDS)
//            .build();
    private ExecutorService mExecutors = Executors.newFixedThreadPool(2);
    public static final int REQUEST_CODE = 100;
    public static final String ORDER_FAIL = "ORDER_FAIL";
    public static final String ORDER_EXCEPTION = "ORDER_EXCEPTION";
    public static final String PAY_FAIL = "PAY_FAIL";
    public static Handler mHandler = new Handler(Looper.getMainLooper());

    private static OkHttpClient getUnsafeOkHttpClient() {
        try {
            final TrustManager[] trustAllCerts = new TrustManager[]{
                    new X509TrustManager() {
                        @Override
                        public void checkClientTrusted(java.security.cert.X509Certificate[] chain, String authType) throws CertificateException {
                        }

                        @Override
                        public void checkServerTrusted(java.security.cert.X509Certificate[] chain, String authType) throws CertificateException {
                        }

                        @Override
                        public java.security.cert.X509Certificate[] getAcceptedIssuers() {
                            return new java.security.cert.X509Certificate[]{};
                        }
                    }
            };

            final SSLContext sslContext = SSLContext.getInstance("SSL");
            sslContext.init(null, trustAllCerts, new java.security.SecureRandom());
            final SSLSocketFactory sslSocketFactory = sslContext.getSocketFactory();

            OkHttpClient.Builder builder = new OkHttpClient.Builder()
                    .connectTimeout(TIME_OUT, TimeUnit.SECONDS)
                    .writeTimeout(TIME_OUT, TimeUnit.SECONDS)
                    .readTimeout(TIME_OUT, TimeUnit.SECONDS);
            builder.sslSocketFactory(sslSocketFactory, (X509TrustManager) trustAllCerts[0]);
            builder.hostnameVerifier((hostname, session) -> true);
            OkHttpClient okHttpClient = builder.build();
            return okHttpClient;
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    /**
     * post请求
     *
     * @param url      请求地址
     * @param params   请求参数
     * @param callback 请求回调
     */
    public <T> void post(String url, Map<String, Object> params, com.bigfun.tm.login.Callback<T> callback) {
        if (TextUtils.isEmpty(url)) throw new IllegalArgumentException("url.length() == 0");
        if (params.isEmpty()) throw new IllegalArgumentException("params.size == 0");
        String json = null;
        try {
            json = EncryptUtil.encryptData(gson.toJson(params));
            Request request = new Request.Builder()
                    .url(url)
                    .post(RequestBody.create(mediaType, json))
                    .build();
            okHttpClient.newCall(request).enqueue(new Callback() {
                @Override
                public void onFailure(Call call, IOException e) {
                    callback.onFail(e.getMessage());
                }

                @Override
                public void onResponse(Call call, Response response) throws IOException {
                    try {
                        if (response.isSuccessful()) {
                            if (response.code() == 200) {
                                if (response.body() != null) {
                                    callback.onResult((T) response.body().string());
                                } else {
                                    callback.onFail(response.code() + "--" + response.message());
                                }
                            } else {
                                callback.onFail(response.code() + "--" + response.message());
                            }
                        } else {
                            callback.onFail(response.code() + "--" + response.message());
                        }
                    } catch (Exception e) {
                        e.printStackTrace();
                        callback.onFail(e.getMessage());
                    }
                }
            });
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    /**
     * 登录
     *
     * @param url
     * @param params
     * @param listener
     */
    public void login(String url, Map<String, Object> params, ResponseListener listener) {
        if (TextUtils.isEmpty(url)) throw new IllegalArgumentException("url.length() == 0");
        if (params.isEmpty()) throw new IllegalArgumentException("params.size == 0");
        String json = null;
        try {
            json = EncryptUtil.encryptData(gson.toJson(params));
            Request request = new Request.Builder()
                    .url(url)
                    .post(RequestBody.create(mediaType, json))
                    .build();
            okHttpClient.newCall(request).enqueue(new Callback() {
                @Override
                public void onFailure(Call call, IOException e) {
                    listener.onFail(e.getMessage());
                }

                @Override
                public void onResponse(Call call, Response response) throws IOException {
                    try {
                        if (response.isSuccessful()) {
                            if (response.code() == 200) {
                                if (response.body() != null) {
                                    LoginBean loginBean =
                                            gson.fromJson(response.body().string(), LoginBean.class);
                                    if (Integer.parseInt(loginBean.getCode()) == 0) {
                                        int loginType = (int) params.get("loginType");
                                        if (loginType == 2) {
                                            String mobile = (String) params.get("mobile");
                                            if (!TextUtils.isEmpty(mobile)) {
                                                SPUtils.getInstance().put(BigFunSDK.mContext, Constant.KEY_LOGIN_PHONE, mobile);
                                            }
                                        }
                                        SPUtils.getInstance().put(
                                                BigFunSDK.mContext,
                                                Constant.KEY_TOKEN,
                                                loginBean.getData().getAccessToken()
                                        );
                                        listener.onSuccess();
                                    } else {
                                        SPUtils.getInstance().remove(BigFunSDK.mContext, "channel");
                                        listener.onFail(loginBean.getMsg());
                                    }
                                } else {
                                    listener.onFail(response.message());
                                }
                            } else {
                                SPUtils.getInstance().remove(BigFunSDK.mContext, "channel");
                                listener.onFail(response.message());
                            }
                        } else {
                            listener.onFail(response.code() + "--" + response.message());
                        }
                    } catch (Exception e) {
                        e.printStackTrace();
                    }
                }
            });
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    private static String mOrderId = "";

    /**
     * 预充值下单
     *
     * @param url
     * @param params
     * @param activity
     * @param listener
     */
    public void payOrder(String url, Map<String, Object> params, Activity activity, ResponseListener listener) {
        String orderId = (String) params.get("orderId");
        if (!TextUtils.isEmpty(mOrderId) && mOrderId.equals(orderId)) {
            return;
        }
        if (NetworkUtils.getNetworkState(BigFunSDK.mContext) == 0) {
            return;
        }
        if (TextUtils.isEmpty(url)) throw new IllegalArgumentException("url.length() == 0");
        if (params.isEmpty()) throw new IllegalArgumentException("params.size == 0");
        mExecutors.execute(new Task(listener, params, activity, url));
        mOrderId = orderId;
    }

    public class Task implements Runnable {

        /**
         * 最大请求次数
         */
        private static final int MAX_TIMES = 3;
        private int mRequestTimes = 0;
        private ResponseListener listener;
        private Map<String, Object> params;
        private Activity activity;

        public Task(ResponseListener listener, Map<String, Object> params, Activity activity, String url) {
            this.listener = listener;
            this.params = params;
            this.activity = activity;
            this.url = url;
        }

        private String url;

        @Override
        public void run() {
            String json = null;
            mRequestTimes = 0;
            while (mRequestTimes < MAX_TIMES) {
                mRequestTimes++;
                try {
                    json = EncryptUtil.encryptData(gson.toJson(params));
                    Request request = new Request.Builder()
                            .url(url)
                            .post(RequestBody.create(mediaType, json))
                            .build();
                    Response response = okHttpClient.newCall(request).execute();
                    if (response.isSuccessful()) {
                        if (response.code() == 200) {
                            if (response.body() != null) {
                                String responseStr = response.body().string();
                                if (!TextUtils.isEmpty(responseStr)) {
                                    PaymentOrderBean bean =
                                            gson.fromJson(
                                                    responseStr,
                                                    PaymentOrderBean.class
                                            );
                                    if (Integer.parseInt(bean.getCode()) == 0) {
                                        if (bean.getData() != null) {
                                            listener.onSuccess();
                                            PayUtils.getInstance().pay(
                                                    bean.getData(),
                                                    activity,
                                                    REQUEST_CODE
                                            );
                                        } else {
                                            report(ORDER_FAIL, bean.getMsg());
                                            listener.onFail(bean.getMsg());
                                        }
                                    } else {
                                        report(ORDER_FAIL, bean.getMsg());
                                        listener.onFail(bean.getMsg());
                                    }
                                } else {
                                    report(ORDER_FAIL, response.message());
                                    listener.onFail(response.message());
                                }
                            } else {
                                report(ORDER_FAIL, response.message());
                                listener.onFail(response.message());
                            }
                        } else {
                            report(ORDER_FAIL, response.message());
                            listener.onFail(response.message());
                        }
                        break;
                    } else {
                        report(ORDER_FAIL, response.message());
                        listener.onFail(response.message());
                    }
                } catch (Exception e) {
                    LogUtils.log("order exception " + e.getMessage());
                    e.printStackTrace();
                    report(ORDER_EXCEPTION, e.getMessage());
                    SystemClock.sleep(100);
                }
            }
        }
    }

    public synchronized void report(String action, String content) {
        mHandler.post(() -> EventManager.getInstance().addEvent(action, content));
    }

    public synchronized void upload(List<EventBean> list) {
        mExecutors.execute(new ReportTask(list));
    }

    public synchronized void query() {
        mHandler.post(() -> EventManager.getInstance().query());
    }
}

File: paytmlibrary/src/main/java/com/bigfun/tm/HttpUtils.kt
package com.bigfun.tm

import android.app.Activity
import com.google.gson.Gson
import com.bigfun.tm.encrypt.EncryptUtil
import com.bigfun.tm.model.LoginBean
import com.bigfun.tm.model.PaymentOrderBean
import com.bigfun.tm.model.SendSmsBean
import okhttp3.*
import java.io.IOException
import java.util.concurrent.TimeUnit


internal class HttpUtils private constructor() {

    private val mediaType = MediaType.get("application/json; charset=utf-8")
    var smsCode = ""
    private val token by lazy {
        SPUtils.instance.get(BigFunSDK.mContext, "accessToken", "") as String
    }

    companion object {
        private const val TIME_OUT = 10L

        @JvmStatic
        val instance by lazy(mode = LazyThreadSafetyMode.SYNCHRONIZED) { HttpUtils() }
    }

    private val gson by lazy(mode = LazyThreadSafetyMode.SYNCHRONIZED) {
        Gson()
    }

    private val okHttpClient by lazy(mode = LazyThreadSafetyMode.SYNCHRONIZED) {
        OkHttpClient.Builder()
            .connectTimeout(TIME_OUT, TimeUnit.SECONDS)
            .writeTimeout(TIME_OUT, TimeUnit.SECONDS)
            .readTimeout(TIME_OUT, TimeUnit.SECONDS)
            .build()
    }

    /**
     * post请求
     * @param url 请求地址
     * @param params 请求参数
     * @param listener 请求回调
     */
//    internal fun <T> post(
//        url: String,
//        params: Map<String, Any>,
//        listener: ResponseListener<T>
//    ) {
//        if (url.isEmpty()) throw IllegalArgumentException("url.length() == 0")
//        if (params.isEmpty()) throw IllegalArgumentException("params.size == 0")
//        val json = EncryptUtil.encryptData(gson.toJson(params))
//        val request = Request.Builder()
//            .url(url)
//            .addHeader(
//                "accessToken",
//                SPUtils.instance.get(PaytmSdk.mContext!!, KEY_TOKEN, "") as String
//            )
//            .post(RequestBody.create(mediaType, json))
//            .build()
//        okHttpClient.newCall(request).enqueue(object : Callback {
//            override fun onFailure(call: Call, e: IOException) {
//                listener.onFail("${e.message}")
//            }
//
//            override fun onResponse(call: Call, response: Response) {
//                try {
//                    if (response.isSuccessful) {
//                        if (response.code() == 200) {
//                            if (response.body() != null) {
//                                val callBackType: Type? = listener.javaClass.genericSuperclass
//                                if (callBackType != null) {
//                                    // 获取泛型类型数组
//                                    val array: Array<Type> =
//                                        (callBackType as ParameterizedType).actualTypeArguments
//                                    val bean: Any =
//                                        gson.fromJson(response.body()!!.string(), array[0])
//                                    listener.onResult(bean as T)
//                                    if (bean is LoginBean) {
//                                        SPUtils.instance.put(
//                                            PaytmSdk.mContext!!,
//                                            KEY_TOKEN,
//                                            bean.data.accessToken
//                                        )
//                                    } else if (bean is SendSmsBean) {
//                                        smsCode = bean.data
//                                    }
//                                } else {
//                                    listener.onFail("请传入泛型参数")
//                                }
//                            } else {
//                                listener.onFail("${response.code()}--${response.message()}")
//                            }
//                        } else {
//                            listener.onFail("${response.code()}--${response.message()}")
//                        }
//                    } else {
//                        listener.onFail("${response.code()}--${response.message()}")
//                    }
//                } catch (e: Exception) {
//                    e.printStackTrace()
//                }
//            }
//        })
//    }

    /**
     * 登录
     */
    fun login(
        url: String,
        params: Map<String, Any>,
        listener: ResponseListener
    ) {
        if (url.isEmpty()) throw IllegalArgumentException("url.length() == 0")
        if (params.isEmpty()) throw IllegalArgumentException("params.size == 0")
        val json = EncryptUtil.encryptData(gson.toJson(params))
        val request = Request.Builder()
            .url(url)
            .addHeader(
                "accessToken",
                SPUtils.instance.get(BigFunSDK.mContext, KEY_TOKEN, "") as String
            )
            .post(RequestBody.create(mediaType, json))
            .build()
        okHttpClient.newCall(request).enqueue(object : Callback {
            override fun onFailure(call: Call, e: IOException) {
                listener.onFail("${e.message}")
            }

            override fun onResponse(call: Call, response: Response) {
                try {
                    if (response.isSuccessful) {
                        if (response.code() == 200) {
                            if (response.body() != null) {
                                val loginBean =
                                    gson.fromJson(response.body()!!.string(), LoginBean::class.java)
                                if (loginBean.code.toInt() == 0) {
                                    SPUtils.instance.put(
                                        BigFunSDK.mContext,
                                        KEY_TOKEN,
                                        loginBean.data.accessToken
                                    )
                                    listener.onSuccess()
                                } else {
                                    listener.onFail(loginBean.msg)
                                }
                            } else {
                                listener.onFail(response.message())
                            }
                        } else {
                            listener.onFail(response.message())
                        }
                    } else {
                        listener.onFail("${response.code()}--${response.message()}")
                    }
                } catch (e: Exception) {
                    e.printStackTrace()
                }
            }

        })
    }

    /**
     * 发送验证码
     */
    fun sendSms(
        url: String,
        params: Map<String, Any>,
        listener: ResponseListener
    ) {
        if (url.isEmpty()) throw IllegalArgumentException("url.length() == 0")
        if (params.isEmpty()) throw IllegalArgumentException("params.size == 0")
        val json = EncryptUtil.encryptData(gson.toJson(params))
        val request = Request.Builder()
            .url(url)
            .addHeader(
                "accessToken",
                SPUtils.instance.get(BigFunSDK.mContext, KEY_TOKEN, "") as String
            )
            .post(RequestBody.create(mediaType, json))
            .build()
        okHttpClient.newCall(request).enqueue(object : Callback {
            override fun onFailure(call: Call, e: IOException) {
                listener.onFail("${e.message}")
            }

            override fun onResponse(call: Call, response: Response) {
                try {
                    if (response.isSuccessful) {
                        if (response.code() == 200) {
                            if (response.body() != null) {
                                val bean =
                                    gson.fromJson(
                                        response.body()!!.string(),
                                        SendSmsBean::class.java
                                    )
                                if (bean.code.toInt() == 0) {
                                    smsCode = bean.data
                                    listener.onSuccess()
                                } else {
                                    listener.onFail(bean.msg)
                                }
                            } else {
                                listener.onFail(response.message())
                            }
                        } else {
                            listener.onFail(response.message())
                        }
                    } else {
                        listener.onFail("${response.code()}--${response.message()}")
                    }
                } catch (e: Exception) {
                    e.printStackTrace()
                }
            }

        })
    }

    /**
     * get请求
     * @param url 请求地址
     * @param params 请求参数
     * @param listener 请求回调
     */
    internal fun <T> get(
        url: String,
        params: Map<String, Any>,
        listener: ResponseListener
    ) {
        try {
            if (url.isEmpty()) throw IllegalArgumentException("url.length() == 0")
            val requestUrl = StringBuffer(url)
            var isFirst = true
            for ((key, value) in params) {
                if (isFirst) {
                    isFirst = false
                    requestUrl.append("?")
                } else {
                    requestUrl.append("&")
                }
                requestUrl.append("$key=$value")
            }
            val request = Request.Builder()
                .url(requestUrl.toString())
                .addHeader("accessToken", token)
                .get()
                .build()
            okHttpClient.newCall(request).enqueue(object : Callback {
                override fun onFailure(call: Call, e: IOException) {
                    listener.onFail("${e.message}")
                }

                override fun onResponse(call: Call, response: Response) {
                }
            })
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    /**
     * 充值下单
     */
    fun paymentOrder(
        url: String,
        params: Map<String, Any>,
        activity: Activity,
        requestCode: Int,
        listener: ResponseListener
    ) {
        if (url.isEmpty()) throw IllegalArgumentException("url.length() == 0")
        if (params.isEmpty()) throw IllegalArgumentException("params.size == 0")
        val json = EncryptUtil.encryptData(gson.toJson(params))
        val request = Request.Builder()
            .url(url)
            .addHeader(
                "accessToken",
                SPUtils.instance.get(BigFunSDK.mContext, KEY_TOKEN, "") as String
            )
            .post(RequestBody.create(mediaType, json))
            .build()
        okHttpClient.newCall(request).enqueue(object : Callback {
            override fun onFailure(call: Call, e: IOException) {
                listener.onFail("${e.message}")
            }

            override fun onResponse(call: Call, response: Response) {
                try {
                    if (response.isSuccessful) {
                        if (response.body() != null) {
                            if (response.code() == 200) {
                                val bean =
                                    gson.fromJson(
                                        response.body()!!.string(),
                                        PaymentOrderBean::class.java
                                    )
                                if (bean.code.toInt() == 0) {
                                    if (bean.data != null) {
                                        listener.onSuccess()
                                        PayUtils.instance.pay(
                                            bean.data,
                                            activity,
                                            requestCode,
                                            params["email"] as String?,
                                            params["mobile"] as String?
                                        )
                                    } else {
                                        listener.onFail(bean.msg)
                                    }
                                } else {
                                    listener.onFail(bean.msg)
                                }
                            } else {
                                listener.onFail(response.message())
                            }
                        } else {
                            listener.onFail(response.message())
                        }
                    } else {
                        listener.onFail(response.message())
                    }
                } catch (e: Exception) {
                    e.printStackTrace()
                }
            }
        })
    }

    /**
     * 获取渠道配置
     */
    fun <T> getChannelConfig(
        url: String,
        params: Map<String, Any>,
        callback: com.bigfun.tm.login.Callback<T>
    ) {
        if (url.isEmpty()) throw IllegalArgumentException("url.length() == 0")
        if (params.isEmpty()) throw IllegalArgumentException("params.size == 0")
        val json = EncryptUtil.encryptData(gson.toJson(params))
        val request = Request.Builder()
            .url(url)
            .addHeader(
                "accessToken",
                SPUtils.instance.get(BigFunSDK.mContext, KEY_TOKEN, "") as String
            )
            .post(RequestBody.create(mediaType, json))
            .build()
        okHttpClient.newCall(request).enqueue(object : Callback {
            override fun onFailure(call: Call, e: IOException) {
                callback.onFail("${e.message}")
            }

            override fun onResponse(call: Call, response: Response) {
                if (response.isSuccessful) {
                    if (response.code() == 200) {
                        if (response.body() != null) {
                            callback.onResult(response.body()!!.string() as T)
                        } else {
                            callback.onFail(response.message())
                        }
                    } else {
                        callback.onFail(response.message())
                    }
                } else {
                    callback.onFail(response.message())
                }
            }
        })
    }
}